app/gui/src/project-view/stores/graph/graphDatabase.ts

```diff
@@ -369,7 +369,13 @@ export class GraphDb {
       }
       const astFields: NodeAstField[] = ['outerAst', 'pattern', 'rootExpr', 'innerExpr']
       astFields.forEach(updateAst)
-      if (oldNode.primarySubject !== primarySubject) node.primarySubject = primarySubject
+      if (
+        !primaryApplicationEquals(
+          oldNode.primaryApplication as PrimaryApplication,
+          primaryApplication,
+        )
+      )
+        node.primaryApplication = primaryApplication
```

Comment:
I think in this approach there's some hazard of future changes causing reactivity bugs. Elsewhere we read the `primaryApplication` field and pass around the resulting object, which has no direct reactive connection to future updates if we replace the object wholesale. It would be more robust to conditionally update each field.

Reply:
Implemented as suggested, but the code is… not nice. Can I improve anything? https://github.com/enso-org/enso/pull/12751/commits/65acd4276160c610b239c95024f256bdc47aeecd

Reply:
To ensure it's deeply reactive, the object would need to never be replaced with a different object--this includes if the value goes to `undefined` and then is set to a new value. So instead of the object being nullable, its fields must be nullable. If the object is always present and its fields are nullable, the update can follow the `if (oldNode.primarySubject !== primarySubject) node.primarySubject = primarySubject` pattern for each field.